### PR TITLE
Enable searching courses by teacher name

### PR DESCRIPTION
**Pull request by HydrogenC**
Enable searching by teacher name on server side, resolves <https://github.com/DanXi-Dev/DanXi/issues/375>.  
In order to make the new code work, there are something changes to do on the database.  
 - Run the following SQL code to create the `teacher` table (for the sake of query performance)  
```sql
CREATE TABLE `teacher`  (
  `id` bigint NOT NULL AUTO_INCREMENT,
  `name` varchar(255),
  PRIMARY KEY (`id`)
)
```
```sql
CREATE TABLE `teacher_course_groups`  (
  `teacher_id` bigint NOT NULL,
  `course_group_id` bigint NOT NULL,
  PRIMARY KEY (`teacher_id`, `course_group_id`),
  FOREIGN KEY (`course_group_id`) REFERENCES `course_group` (`id`),
  FOREIGN KEY (`teacher_id`) REFERENCES `teacher` (`id`)
)
```
 - Run this go script to insert data into table `teacher` and `teacher_courses` from `course`  
Link: <https://github.com/HydrogenC/backend/blob/teacher-name/danke_utils/teacher_table.go>

### CODE REVIEW

**Comment by HydrogenC** [comment]
Also included a fix for #134.

File: danke_utils/teacher_table.go [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+	for {
+		query := DB.Table("course").Limit(BatchSize).Offset(offset)
+		query.Find(&queryResult)
+		offset += len(queryResult)
+
+		if len(queryResult) == 0 {
+			// Finished iterating
+			break
+		}
+
```
Comment (JingYiJun) [inline danke_utils/teacher_table.go:52]: 这里可以用 db.FindInBatches ，避免手动维护 offset

File: danke_utils/teacher_table.go [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+		for _, course := range queryResult {
+			teacherList := strings.Split(course.Teachers, ",")
+			for _, name := range teacherList {
+				courseList, found := dataMap[name]
+				if found {
+					dataMap[name] = append(courseList, course.CourseGroupID)
```
Comment (JingYiJun) [inline danke_utils/teacher_table.go:58]: 这里 CourseGroupID 是否会重复，是否需要做一下去重

**Comment by HydrogenC** [comment]
Resolved.

**Comment by HydrogenC** [comment]
Do I have to separate a new API for teacher name searching and use `db.Table("teacher").Preload("CourseGroups").Find(...)` to find related course groups?

**Comment by JingYiJun** [comment]
> Do I have to separate a new API for teacher name searching and use `db.Table("teacher").Preload("CourseGroups").Find(...)` to find related course groups?

或许这样是更好的，不过需要前端适配两种不同的搜索模式。现在这样的策略对于用户来说，搜索 Teacher 但是显示 CourseGroup 感觉比较奇怪😂

File: danke/api/course_group.go [hunk 3]
```diff
@@ -148,7 +149,12 @@ func SearchCourseGroupV3(c *fiber.Ctx) (err error) {
 	if CourseCodeRegexp.MatchString(query) {
 		querySet = querySet.Where("code LIKE ?", query+"%")
 	} else {
-		querySet = querySet.Where("name LIKE ?", "%"+query+"%")
+		queryWord :="%"+query+"%"
+		querySet = querySet.
+			Joins("JOIN teacher_course_groups tcg ON course_group.id = tcg.course_group_id").
+			Joins("JOIN teacher t ON tc.teacher_id = t.id").
+			Where("t.name like ? OR course_group.name LIKE ?", queryWord, queryWord).
+			Group("id")
```
Comment (JingYiJun) [inline danke/api/course_group.go:157]: 这里关联表名称是 teacher_course_groups，在 many2many 声明的 是 teacher_course_link，在 GenerateTeacherTable 中是 Table("teacher_courses")，这个是不是需要明确一下？统一用 teacher_course_groups ？作为关联？

Comment (HydrogenC) [reply danke/api/course_group.go:157]: My fault, I just copied without checking and modifying.

**Comment by HydrogenC** [comment]
> > Do I have to separate a new API for teacher name searching and use `db.Table("teacher").Preload("CourseGroups").Find(...)` to find related course groups?
> 
> 或许这样是更好的，不过需要前端适配两种不同的搜索模式。现在这样的策略对于用户来说，搜索 Teacher 但是显示 CourseGroup 感觉比较奇怪😂

Is there an advantage for using `Preload` over `Join`? Should I modify my current implementation then?

**Comment by JingYiJun** [comment]
> > > Do I have to separate a new API for teacher name searching and use `db.Table("teacher").Preload("CourseGroups").Find(...)` to find related course groups?
> > 
> > 
> > 或许这样是更好的，不过需要前端适配两种不同的搜索模式。现在这样的策略对于用户来说，搜索 Teacher 但是显示 CourseGroup 感觉比较奇怪😂
> 
> Is there an advantage for using `Preload` over `Join`? Should I modify my current implementation then?

Maybe in this situation `Join` is better since `Join` means only one load from database while `Preload` implies 3 loads, and there is no need to get the intermediate information of `teachers`. If there you add a new api to load the teachers user search, `Preload` may better, but in this pr you've done enough on changing SearchCourseGroups.

**Review by JingYiJun — APPROVED** [review]

**Pull request merged**